# A6 : Naive RAG vs Contextual Retrieval

Since my studentID is 124371 I am using chapter 11. The link to the doc is provided [here](https://web.stanford.edu/~jurafsky/slp3/11.pdf) 

In [1]:
from langchain_community.document_loaders import PyPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter

In [2]:
loader = PyPDFLoader("11.pdf")
raw_docs = loader.load()

In [3]:
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=50,
    length_function=len,
    separators=["\n\n", "\n", " ", ""]
)

In [4]:
documents = text_splitter.split_documents(raw_docs)

In [5]:
print(f"Total chunks created: {len(documents)}")
print(f"Sample content from first chunk: {documents[0].page_content[:300]}")

Total chunks created: 163
Sample content from first chunk: Speech and Language Processing. Daniel Jurafsky & James H. Martin. Copyright © 2026. All
rights reserved. Draft of January 6, 2026.
CHAPTER
11
Information Retrieval and
Retrieval-Augmented Generation
On two occasions I have been asked,—“Pray, Mr. Babbage, if you put into
the machine wrong figures, wi


## Creating 20 Question Answer pairs from the doc 

In [6]:
pair_20 = [
  {
    "question": "What is the primary goal of Information Retrieval (IR)?",
    "ground_truth_answer": "To find documents that are relevant to a user’s information need from a large collection of unstructured text.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Define 'Retrieval-Augmented Generation' (RAG).",
    "ground_truth_answer": "An architecture that combines a pre-trained LLM with an external retrieval system to provide contextually relevant and factual information.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is an 'Inverted Index'?",
    "ground_truth_answer": "A data structure that maps each unique term to a list of the documents (postings) in which it appears.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Explain the term 'Term Frequency' (TF).",
    "ground_truth_answer": "A measure of how often a specific word appears in a document, often normalized to prevent bias toward longer documents.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What does 'IDF' stand for and what is its purpose?",
    "ground_truth_answer": "Inverse Document Frequency; it diminishes the weight of terms that occur very frequently across the entire corpus and increases the weight of rare terms.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "How is the BM25 algorithm different from basic TF-IDF?",
    "ground_truth_answer": "BM25 improves on TF-IDF by incorporating document length normalization and term frequency saturation.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is the 'Vector Space Model'?",
    "ground_truth_answer": "A model where documents and queries are represented as points (vectors) in a high-dimensional space based on term weights.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Define 'Cosine Similarity' in the context of IR.",
    "ground_truth_answer": "A metric used to measure the similarity between two vectors by calculating the cosine of the angle between them.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What are 'Dense Vectors' (Embeddings)?",
    "ground_truth_answer": "Continuous, low-dimensional representations of text generated by neural models that capture semantic meaning rather than just word overlap.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is the role of the 'Retriever' in a RAG system?",
    "ground_truth_answer": "To search the knowledge base and select the top k document chunks most relevant to the user's input query.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is the role of the 'Generator' in a RAG system?",
    "ground_truth_answer": "To take the user query and the retrieved context chunks to produce a coherent, natural language response.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Explain 'Precision' in IR evaluation.",
    "ground_truth_answer": "The fraction of retrieved documents that are actually relevant to the user's query.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Explain 'Recall' in IR evaluation.",
    "ground_truth_answer": "The fraction of all relevant documents in the collection that were successfully retrieved by the system.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is the 'F-measure'?",
    "ground_truth_answer": "A single score that represents the harmonic mean of Precision and Recall, balancing both metrics.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is 'Mean Reciprocal Rank' (MRR)?",
    "ground_truth_answer": "An evaluation metric that looks at where the first relevant document appears in the ranked list of results.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Define 'Nearest Neighbor Search' (NNS).",
    "ground_truth_answer": "The process of finding the vectors in a database that are closest to a query vector, often used in vector databases for RAG.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is 'Knowledge Grounding'?",
    "ground_truth_answer": "The process of ensuring an LLM's output is based on provided factual evidence (the retrieved documents) to reduce hallucinations.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Explain 'Bi-Encoders' in dense retrieval.",
    "ground_truth_answer": "A model architecture where the query and the document are encoded separately into vectors before being compared.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "What is the 'Parametric Knowledge' of an LLM?",
    "ground_truth_answer": "The information stored within the model's weights during its initial training phase.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  },
  {
    "question": "Why is 'Chunking' necessary for RAG?",
    "ground_truth_answer": "Because LLMs have limited context windows and smaller chunks allow for more precise and diverse retrieval of relevant information.",
    "naive_rag_answer": "",
    "contextual_retrieval_answer": ""
  }
]

### Naive RAG

In [7]:
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma
import torch
import torch.nn as nn

embeddings = HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")
# Test it
query_result = embeddings.embed_query("What is RAG?")
print(len(query_result)) 


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


384


In [8]:
vector_db = Chroma.from_documents(
    documents=documents, 
    embedding=embeddings,
    persist_directory="./chroma_db"
)

In [9]:
retriever = vector_db.as_retriever(search_kwargs={"k": 3})

In [10]:
import getpass
import os

os.environ["GROQ_API_KEY"] = getpass.getpass("Enter your Grok API key: ")

Enter your Grok API key:  ········


In [11]:
from langchain_openai import ChatOpenAI
from langchain_classic.chains import RetrievalQA

#initialize the llm
from langchain_groq import ChatGroq

llm = ChatGroq(
    model_name="llama-3.1-8b-instant", 
    temperature=0
)

#create the chain
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=retriever,
    return_source_documents=True
)

In [12]:
# Example query from your JSON list
query = "What is an Inverted Index?"
response = qa_chain.invoke({"query": query})

print(f"Answer: {response['result']}")
# You can also see which chunks were used:
print(f"Source: {response['source_documents'][0].metadata}")

Answer: According to the provided historical notes, an Inverted Index is a storage mechanism for sparse retrieval that makes it very efficient to find documents that have a particular word.
Source: {'total_pages': 25, 'trapped': '/False', 'title': '', 'source': '11.pdf', 'author': '', 'ptex.fullbanner': 'This is pdfTeX, Version 3.14159265-2.6-1.40.21 (TeX Live 2020) kpathsea version 6.3.2', 'producer': 'pdfTeX-1.40.21', 'creator': 'LaTeX with hyperref', 'page_label': '21', 'moddate': '2026-01-06T08:23:25-08:00', 'page': 20, 'creationdate': '2026-01-06T08:23:25-08:00', 'subject': '', 'keywords': ''}


### Contextual

In [13]:
import asyncio
from langchain_groq import ChatGroq
from langchain_core.documents import Document
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma

In [14]:
async def enrich_chunk(chunk_text: str, full_doc_text: str, title: str) -> str:
    """Generates a situational prefix for a chunk."""
    prompt = f"""
    <document>
    {full_doc_text[:4000]} 
    </document>
    
    Here is a specific chunk from the document:
    <chunk>
    {chunk_text}
    </chunk>
    
    Provide a 1-sentence context situating this chunk within the overall document.
    Format: "This chunk from [title] discusses [explanation]."
    """
    
    response = await llm.ainvoke(prompt)
    context_prefix = response.content.strip()
    return f"{context_prefix}\n\n{chunk_text}"

In [15]:
async def process_all_chunks(chunks, full_text, title):
    enriched_docs = []
    print(f"Enriching {len(chunks)} chunks... please wait.")
    
    for chunk in chunks:
        # Create the contextualized text
        contextualized_text = await enrich_chunk(chunk.page_content, full_text, title)
        
        # Preserve original metadata but update the content
        new_doc = Document(
            page_content=contextualized_text,
            metadata=chunk.metadata
        )
        enriched_docs.append(new_doc)
    
    return enriched_docs

In [16]:
full_text = "\n\n".join([doc.page_content for doc in raw_docs])

In [17]:
enriched_chunks = await process_all_chunks(
    chunks=documents, 
    full_text=full_text, 
    title="Chapter 11: Information Retrieval"
)

Enriching 163 chunks... please wait.


In [18]:
embeddings_gpu = HuggingFaceEmbeddings(
    model_name="all-MiniLM-L6-v2", 
    model_kwargs={'device': 'cuda'}
)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [19]:
contextual_db = Chroma.from_documents(
    documents=enriched_chunks,
    embedding=embeddings_gpu,
    persist_directory="./chroma_contextual"
)

In [20]:
contextual_retriever = contextual_db.as_retriever(search_kwargs={"k": 3})

contextual_qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=contextual_retriever,
    return_source_documents=True
)

In [21]:
query = "What is an inverted index?"

#naive retrieval
naive_results = vector_db.similarity_search(query, k=1)
print(f"NAIVE MATCH: {naive_results[0].page_content[:200]}...")

#contextual retrieval
contextual_results = contextual_db.similarity_search(query, k=1)
print(f"CONTEXTUAL MATCH: {contextual_results[0].page_content[:200]}...")

NAIVE MATCH: HISTORICAL NOTES 21
• The inverted index is a storage mechanism for sparse retrieval that makes it
very efficient to find documents that have a particular word.
• In dense retrieval techniques, document...
CONTEXTUAL MATCH: This chunk from "Speech and Language Processing" discusses a simple example of an inverted index, a data structure used in information retrieval to efficiently store and retrieve terms from a collecti...


In [22]:
import json


contextual_retriever = contextual_db.as_retriever(search_kwargs={"k": 3})

contextual_qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=contextual_retriever,
    return_source_documents=True
)

print("Starting evaluation loop for 20 questions...")

for i, item in enumerate(pair_20):
    question = item["question"]
    print(f"[{i+1}/20] Processing: {question}")
    
    try:
        naive_response = qa_chain.invoke({"query": question})
        item["naive_rag_answer"] = naive_response["result"]
        
        contextual_response = contextual_qa_chain.invoke({"query": question})
        item["contextual_retrieval_answer"] = contextual_response["result"]
    except Exception as e:
        print(f"Error processing question {i+1}: {e}")
        item["naive_rag_answer"] = "ERROR"
        item["contextual_retrieval_answer"] = "ERROR"

with open('rag_results.json', 'w') as f:
    json.dump(pair_20, f, indent=4)

print("Done! All answers generated and saved to 'rag_results.json'.")



Starting evaluation loop for 20 questions...
[1/20] Processing: What is the primary goal of Information Retrieval (IR)?
[2/20] Processing: Define 'Retrieval-Augmented Generation' (RAG).
[3/20] Processing: What is an 'Inverted Index'?
[4/20] Processing: Explain the term 'Term Frequency' (TF).
[5/20] Processing: What does 'IDF' stand for and what is its purpose?
[6/20] Processing: How is the BM25 algorithm different from basic TF-IDF?
[7/20] Processing: What is the 'Vector Space Model'?
[8/20] Processing: Define 'Cosine Similarity' in the context of IR.
[9/20] Processing: What are 'Dense Vectors' (Embeddings)?
[10/20] Processing: What is the role of the 'Retriever' in a RAG system?
[11/20] Processing: What is the role of the 'Generator' in a RAG system?
[12/20] Processing: Explain 'Precision' in IR evaluation.
[13/20] Processing: Explain 'Recall' in IR evaluation.
[14/20] Processing: What is the 'F-measure'?
[15/20] Processing: What is 'Mean Reciprocal Rank' (MRR)?
[16/20] Processing: De

In [24]:
# Preview the  result to verify
print("\nSample Result:")
print(f"Question: {pair_20[10]['question']}")
print(f"Naive: {pair_20[0]['naive_rag_answer'][:100]}...")
print(f"Contextual: {pair_20[0]['contextual_retrieval_answer'][:100]}...")


Sample Result:
Question: What is the role of the 'Generator' in a RAG system?
Naive: The primary goal of Information Retrieval (IR) is to give a sufficient overview of IR to see its app...
Contextual: The primary goal of Information Retrieval (IR) is to retrieve various media based on user informatio...


In [27]:
import json
import pandas as pd
from rouge_score import rouge_scorer


with open('rag_results.json', 'r') as f:
    data = json.load(f)

scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

results_table = []

for entry in data:
    question = entry['question']
    gt = entry['ground_truth_answer']
    
    for model_name, column in [("Naive RAG", "naive_rag_answer"), ("Contextual Retrieval", "contextual_retrieval_answer")]:
        candidate = entry[column]
        
        # Handle "I don't know" cases 
        scores = scorer.score(gt, candidate)
        
        results_table.append({
            "Question": (question[:45] + '...') if len(question) > 45 else question,
            "Model": model_name,
            "ROUGE-1 (F1)": round(scores['rouge1'].fmeasure, 4),
            "ROUGE-2 (F1)": round(scores['rouge2'].fmeasure, 4),
            "ROUGE-L (F1)": round(scores['rougeL'].fmeasure, 4)
        })

df = pd.DataFrame(results_table)
print(df.to_string(index=False))

                                        Question                Model  ROUGE-1 (F1)  ROUGE-2 (F1)  ROUGE-L (F1)
What is the primary goal of Information Retri...            Naive RAG        0.3478        0.0455        0.2174
What is the primary goal of Information Retri... Contextual Retrieval        0.2778        0.0588        0.2222
Define 'Retrieval-Augmented Generation' (RAG)...            Naive RAG        0.1692        0.0469        0.1538
Define 'Retrieval-Augmented Generation' (RAG)... Contextual Retrieval        0.1379        0.0349        0.1379
                    What is an 'Inverted Index'?            Naive RAG        0.2917        0.0000        0.1667
                    What is an 'Inverted Index'? Contextual Retrieval        0.2222        0.0845        0.1806
         Explain the term 'Term Frequency' (TF).            Naive RAG        0.1657        0.0719        0.1420
         Explain the term 'Term Frequency' (TF). Contextual Retrieval        0.1165        0.0490       

## Understanding the results

<b>Low Overall Scores:</b> Most ROUGE-1 scores hover between 0.10 and 0.35, while ROUGE-2 (bigram overlap) is frequently below 0.10. This suggests that while the models capture the general concepts, they do not perfectly mimic the concise phrasing of the ground truth.

<b>Naive RAG’s Edge in Precision:</b> In several instances, such as the "Primary Goal of IR" and "Role of the Generator," Naive RAG achieves higher ROUGE-1 and ROUGE-L scores than Contextual Retrieval. This often happens when the Naive model provides a shorter answer that aligns more closely with the reference length.

Some findings interpreted from the experiment are :

1. ROUGE-2 (F1) "Zero" Scores: There are multiple instances (e.g., "Inverted Index" and "Knowledge Grounding" for Naive RAG) where the ROUGE-2 score is 0.0000. This indicates that while individual words might match (ROUGE-1), the model failed to capture any exact two-word sequences from the ground truth.

2. Contextual Retrieval Successes: Contextual Retrieval significantly outperforms Naive RAG on the "IDF" question, nearly doubling the ROUGE-1 score (0.2969 vs 0.1758) and ROUGE-2 score (0.1587 vs 0.0444). This suggests that for certain technical definitions, the contextual approach retrieved more precise terminology.

3. Model Failures (MRR and Knowledge Grounding): For "Mean Reciprocal Rank (MRR)" and "Knowledge Grounding," the Contextual Retrieval model scored 0.0000 across all metrics. Cross-referencing this with the raw data shows that the model responded with "I don't know," leading to zero overlap with the factual ground truth.